File: day4/t1.py
def is_valid(line):
    d = {}
    for entry in line.split(" "):
        if d.get(entry, False):
            return False
        else:
            d[entry] = 1
    return True


def main():
    with open('input.txt', 'r') as f:
        data = [d.strip() for d in f.readlines()]
    answers = []
    for line in data:
        answers.append(is_valid(line))

    print("True:", answers.count(True))
    print("False:", answers.count(False))


main()

File: day7/t1.rb
data = IO.readlines("input")
nodes = Hash.new {|hash, key| hash[key] = {children: [], parent: nil}}
data.map! {|line| line.delete(',').split(' ')}
data.each do |line|
    parent = nodes[line[0]]
    # this node has children
    if line.length > 3
        line.drop(3). each do |child|
            child = nodes[child]
            child[:parent] = parent
            parent[:children].push(child)
         end
    end
end
print nodes.find {|k, v| v[:parent].nil?}.first